### PR TITLE
Shirley import url schemes

### PR DESCRIPTION
**Pull request by shirleymango**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature!

**What is the current behavior?** (You can also link to an open issue here)
* Can only import from Google Sheets

**What is the new behavior (if this is a feature change)?**
* User clicks on their friend's link
* Floofcards app opens up
* Sends user to preview page
* User can import their friend's cards

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

**Video Demos**:

https://user-images.githubusercontent.com/44847817/182922433-31ce2044-eb1a-4c8a-b734-3d4506c5b489.mov

**Lessons Learned**:
* Using URL schemes

### CODE REVIEW

File: metau-capstone/Delegates/AppDelegate.m [hunk 1]
```diff
@@ -33,6 +33,10 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
     return YES;
 }
 
+- (BOOL)application:(UIApplication *)app openURL:(NSURL *)url options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options {
+    NSLog(@"open url through custom url scheme");
```
Comment (ebeweber-fb) [inline metau-capstone/Delegates/AppDelegate.m:37]: Next week as we continue to do polish we can remove some of these logs and clean up our code a bit.

File: metau-capstone/Delegates/SceneDelegate.m [hunk 2]
```diff
@@ -22,14 +25,30 @@ - (void)scene:(UIScene *)scene willConnectToSession:(UISceneSession *)session op
     PFUser *user = [PFUser currentUser];
     if (user != nil) {
         NSLog(@"Welcome back %@ 😀", user.username);
-
-        // Load Chat view controller and set as root view controller
```
Comment (ebeweber-fb) [inline metau-capstone/Delegates/SceneDelegate.m:26]: Curious, why did you remove this comment?

File: metau-capstone/Delegates/SceneDelegate.m [hunk 3]
```diff
@@ -22,14 +25,30 @@ - (void)scene:(UIScene *)scene willConnectToSession:(UISceneSession *)session op
...
+    [query whereKey:@"userID" equalTo:userID];
+    [query findObjectsInBackgroundWithBlock:^(NSArray<Flashcard *> *cards, NSError * _Nullable error) {
+        if (!error) {
+            // Create array of Preview Cards
+            [APIManager shared].previewFlashcards = [PreviewFlashcard createCardsFromArray:cards];
+            // Set view controller to preview view controller
+            UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
+            UINavigationController *previewNavigationController = [storyboard instantiateViewControllerWithIdentifier:@"PreviewNavigationController"];
+            self.window.rootViewController = previewNavigationController;
+        }
```
Comment (ebeweber-fb) [inline metau-capstone/Delegates/SceneDelegate.m:49]: Good polish will be handling the error case!

What are you think about when an error is returned?

File: metau-capstone/Delegates/SceneDelegate.m [hunk 4]
```diff
@@ -22,14 +25,30 @@ - (void)scene:(UIScene *)scene willConnectToSession:(UISceneSession *)session op
...
         UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
         UIViewController *tabBarNavigationController = [storyboard instantiateViewControllerWithIdentifier:@"tabBarController"];
         self.window.rootViewController = tabBarNavigationController;
     }
 }
 
+- (void)scene:(UIScene *)scene openURLContexts:(NSSet<UIOpenURLContext *> *)URLContexts {
+    NSURL *url = [URLContexts allObjects][0].URL;
+    NSArray *queryPair = [url.query componentsSeparatedByString:@"="];
+    NSString *userID = queryPair[1];
```
Comment (ebeweber-fb) [inline metau-capstone/Delegates/SceneDelegate.m:37]: In the future if you chose to add other url parameters (say custom decks, or a key to verify that you have permission to download someone's cards, etc) splitting the query string like this might always work.

Say we have `?key=123&userId=456` and `?userId=456&key=123`. We want to be able to deconstruct this and get the userId, but these hard rules (indexing first after the split) might not work consistently.

**Review by ebeweber-fb — COMMENTED** [review]
So rad!

Some polish comments, but way to get this working!

File: metau-capstone/Delegates/SceneDelegate.m (hunk 2 above)
Comment (shirleymango) [reply metau-capstone/Delegates/SceneDelegate.m:26]: I think I copied some code from Parse chat project and carried that comment over. I can add a comment on loading the study view controller!

File: metau-capstone/Delegates/SceneDelegate.m (hunk 4 above)
Comment (shirleymango) [reply metau-capstone/Delegates/SceneDelegate.m:37]: Okay, I refactored the code to create a queryStringDictionary. Instead of getting the userID with a hard-coded index number, I get the value correlated with the @"userID" key.
